### PR TITLE
move particle spawners to env

### PR DESCRIPTION
**Pull request by obneq**
move m_particlespawner_ids and related housekeeping to env.
add code to handle expiration of particle ids for expiring spawners.

here is some code to test with:
https://gist.github.com/obneq/e46553eb518f3ef79ac2

note ever increasing ids for expiring spawners.

### CODE REVIEW

**Comment by est31** [comment]
you should:

1. fix the style when you touch code.
1. squash your commits into a single one.

File: src/environment.cpp [hunk 1]
```diff
@@ -1256,6 +1256,56 @@ void ServerEnvironment::step(float dtime)
...
+	for(;;) // look for unused particlespawner id
+	{
+		id++;
+		std::map<u32, float>::iterator f;
+		f = m_particle_spawners.find(id);
+
+		if(f == m_particle_spawners.end())
+		{
+			std::pair<u32, float> ps = std::make_pair(id, exptime);
+			m_particle_spawners.insert(ps);
```
Comment (est31) [inline src/environment.cpp]: `m_particle_spawners[id] = exptime;` is more readable.

**Comment by est31** [comment]
:+1:

File: src/environment.cpp [hunk 2]
```diff
@@ -1256,6 +1256,50 @@ void ServerEnvironment::step(float dtime)
...
+
+	/*
+		Manage particle spawner expiration
+	*/
+	if (m_particle_management_interval.step(dtime, 1.0)) {
+		for(std::map<u32, float>::iterator i =
+		    m_particle_spawners.begin();
+		    i != m_particle_spawners.end();) {
+			//non expiring spawners
+			if (i->second <= -1.f) {
```
Comment (OttoLidenbrock) [inline src/environment.cpp]: I think it might be safer if this was < 0.0f

File: src/environment.cpp [hunk 3]
```diff
@@ -1256,6 +1256,50 @@ void ServerEnvironment::step(float dtime)
 		*/
 		removeRemovedObjects();
 	}
+
+	/*
+		Manage particle spawner expiration
+	*/
+	if (m_particle_management_interval.step(dtime, 1.0)) {
```
Comment (Zeno-) [inline src/environment.cpp]: Anything wrong with:

```
	if (m_particle_management_interval.step(dtime, 1.0)) {
		for (std::map<u32, float>::iterator it = m_particle_spawners.begin();
				it != m_particle_spawners.end(); it++) {
			if (it > 0.0f) {	// spawners that expire
				it->second -= 1.0;
				if (it->second < 0.0f)
					m_particle_spawners.erase(it);
			}
		}
	}
```

File: src/environment.cpp [hunk 4]
```diff
@@ -1256,6 +1256,50 @@ void ServerEnvironment::step(float dtime)
...
+			else
+				i++;
+		}
+	}
+}
+
+u32 ServerEnvironment::addParticleSpawner(float exptime)
+{
+	// Timers with lifetime 0 do not expire
+	float time = exptime > 0.f ? exptime : -1.f;
```
Comment (Zeno-) [inline src/environment.cpp]: This would be better with the scope limited (e.g. just before line 1293)

**Comment by obneq** [comment]
@Zeno- sorry i dont understand what you mean. but please do feel free to treat all my pulls as bug reports and fix it better! I just want to help despite lacking skills.

**Comment by kilbith** [comment]
[Needs second review] here please.

**Comment by neoascetic** [comment]
Bump?

**Comment by sofar** [comment]
Needs rebase

**Comment by paramat** [comment]
obneq might open a fresh PR for this to avoid a diificult rebase.

**Comment by sofar** [comment]
@obneq please do. I looked at rebasing and it confuses even me :)

**Comment by paramat** [comment]
@obneq i've asked Zeno- to complete this since he reviewed and commented, and it looks like he might do this, so you may not have to bother. I'll update this thread with news.

**Comment by Zeno-** [comment]
A patch file to rebase is attached. Please check it carefully

[rebase_2602.txt](https://github.com/minetest/minetest/files/234455/rebase_2602.txt)

**Comment by Zeno-** [comment]
https://github.com/Zeno-/minetest/commits/obneq

File: src/environment.cpp (hunk 3 above)
Comment (kwolekr) [inline src/environment.cpp]: it gets invalidated after .erase

Comment (Zeno-) [inline src/environment.cpp]: Hmm, yes. Isn't that the case with the original code as well in that case (line 1275)?

File: src/environment.cpp [hunk 5]
```diff
@@ -1256,6 +1256,50 @@ void ServerEnvironment::step(float dtime)
...
+			if (i->second <= -1.f) {
+				i++;
+				continue;
+			}
+
+			i->second -= 1.0;
+			if (i->second < 0.f)
+				m_particle_spawners.erase(i++);
+			else
+				i++;
```
Comment (obneq) [inline src/environment.cpp]: so after reading up on this i wonder if this is correct, ++i is suggested in order not to skip any elements. also i think zenos 2nd version above is clearer than mine but has the same postfix increment, so im commenting here

Comment (kwolekr) [inline src/environment.cpp]: D'oh... see the discussion about this in the channel about this...

Both your version and Zeno's version have 2 rather serious bugs relating to iterators not getting incremented and problems if floating point arithmetic *doesn't* result in any floating point error.

File: src/environment.cpp (hunk 4 above)
Comment (kwolekr) [inline src/environment.cpp]: Also would be nice to not shadow the function `time`

**Comment by Zeno-** [comment]
Rebased with bug fix (refer to IRC logs):

[rebase_2602_v2.txt](https://github.com/minetest/minetest/files/236123/rebase_2602_v2.txt)

**Comment by Zeno-** [comment]
Rebased in https://github.com/minetest/minetest/pull/4038

**Pull request closed without merging**